### PR TITLE
[SPARK-16121] ListingFileCatalog does not list in parallel anymore

### PR DESCRIPTION
**Pull request by yhuai**
## What changes were proposed in this pull request?

Seems the fix of SPARK-14959 breaks the parallel partitioning discovery. This PR fixes the problem
## How was this patch tested?

Tested manually. (This PR also adds a proper test for SPARK-14959)

### CODE REVIEW

**Comment by liancheng** [comment]
LGTM, merging to master and branch-2.0.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ListingFileCatalog.scala [hunk 1]
```diff
@@ -73,21 +73,67 @@ class ListingFileCatalog(
...
+
+        childStatuses.map {
+          case f: LocatedFileStatus => f
+
+          // NOTE:
+          //
+          // - Although S3/S3A/S3N file system can be quite slow for remote file metadata
+          //   operations, calling `getFileBlockLocations` does no harm here since these file system
+          //   implementations don't actually issue RPC for this method.
+          //
```
Comment (steveloughran) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ListingFileCatalog.scala]: while you are correct about s3a, the openstack swift client does do RPC here, as swift instances can exhibit locality, as does Azure. This comment isn't actually correct.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ListingFileCatalog.scala]: Thank you for pointing out that. Can you also provide some pointers? We can put those pointers in the comment.

Comment (steveloughran) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ListingFileCatalog.scala]: Maybe just say "Location aware object stores may trigger queries of the object store here" without being specific about the store. As an example: `org.apache.hadoop.fs.swift.snative.SwiftNativeFileSystem` does this.

**Pull request closed without merging**